src/aspire/source/__init__.py

```diff
@@ -1,57 +1,98 @@
-from aspire.estimation.noise import WhiteNoiseEstimator
 from aspire.utils import ensure
+from aspire.utils.filters import MultiplicativeFilter
 from aspire.utils.coor_trans import grid_2d
 from aspire.io.starfile import StarFile, StarFileBlock
+from aspire.source.xform import Pipeline, DownSample, Filter
+from aspire.source.xform import Multiply, Shift, Filter, IndexedXform
 
 logger = logging.getLogger(__name__)
 
 
 class ImageSource:
 
-    # ----------------------------------------------------
-    # Class Attributes that can be overridden by subclasses
-    # ----------------------------------------------------
-
-    # Optional renaming of metadata fields, used
-    # These are used ONLY during serialization/deserialization (to/from StarFiles, for example).
-    _metadata_aliases = {
-        '_image_name':  '_rlnImageName',
-        '_offset_x':    '_rlnOriginX',
-        '_offset_y':    '_rlnOriginY',
-        '_state':       '_rlnClassNumber',
-        '_angle_0':     '_rlnAngleRot',
-        '_angle_1':     '_rlnAngleTilt',
-        '_angle_2':     '_rlnAnglePsi',
-        '_amplitude':   '_amplitude',
-        '_voltage':     '_rlnVoltage',
-        '_defocus_u':   '_rlnDefocusU',
-        '_defocus_v':   '_rlnDefocusV',
-        '_defocus_ang': '_rlnDefocusAngle',
-        '_Cs':          '_rlnSphericalAberration',
-        '_alpha':       '_rlnAmplitudeContrast'
+    """
+    When creating an ImageSource object, a 'metadata' table holds metadata information about all images in the
+    ImageSource. The number of rows in this metadata table will equal the total no. of images supported by this
+    ImageSource (available as the 'n' attribute), though reading/writing of images is usually done in chunks.
+
+    This metadata table is implemented as a pandas DataFrame.
+
+    The 'values' in this metadata table are usually primitive types (floats/ints/strings) that are suitable
+    for being read from STAR files, and being written to STAR files. The columns corresponding to these fields
+    begin with a single underscore '_'.
+
+    In addition, the metadata table may also contain references to Python objects.
+    'Filter' objects, for example, are stored in this metadata table as references  to unique Filter objects that
+    correspond to images in this ImageSource. Several rows of metadata may end up containing a reference to a small
+    handful of unique Filter objects, depending on the values found in other columns (identical Filter
+    objects, depending on unique CTF values found for _rlnDefocusU/_rlnDefocusV etc.
+    """
+
+    """
+    The metadata_fields dictionary below specifies default data types of certain key fields used in the codebase.
+    The STAR file used to initialize subclasses of ImageSource may well contain other columns not found below; these
+    additional columns are available when read, and they default to the pandas data type 'object'.
+    """
+    metadata_fields = {
```

Comment:
What's the motivation for using the RELION column names in our code? I thought there was a translation layer at some point where these were changed into simpler names.

Reply:
The translation layer was causing a lot of noise (lots of forward and reverse dictionaries applied at different places) for very little gains. It's largely irrelevant since access to key properties (offsets, amplitudes) is through properties which look up the metadata field.

I propose we add back a translation layer if we ever decide to support multiple serialization formats to/from starfile (in addition to Relion).

Reply:
> I propose we add back a translation layer if we ever decide to support multiple serialization formats to/from starfile (in addition to Relion).

Ok sounds good.